### PR TITLE
LPD-44400 Changing scheduled startDate when cronexpression is set.

### PR DESCRIPTION
**Pull request by ZakiMiklos**
[LPD-44400](https://liferay.atlassian.net/browse/LPD-44400)

### CODE REVIEW

**Comment by ZakiMiklos** [comment]
ci:test:sf

**Comment by ZakiMiklos** [comment]
ci:test:relevant

File: modules/apps/dispatch/dispatch-service/src/main/java/com/liferay/dispatch/internal/helper/DispatchTriggerHelper.java [hunk 1]
```diff
@@ -37,9 +41,30 @@ public void addSchedulerJob(
...
+		if ((startDate == null) || startDate.before(now)) {
+			startDate = now;
+		}
+
+		if (dispatchTrigger.getCronExpression() != null) {
+			try {
+				CronExpression ce = new CronExpression(
+					dispatchTrigger.getCronExpression());
+
+				startDate = ce.getNextValidTimeAfter(startDate);
```
Comment (ccorreagg) [inline modules/apps/dispatch/dispatch-service/src/main/java/com/liferay/dispatch/internal/helper/DispatchTriggerHelper.java:56]: I am not sure if importing the quartz library is a good idea, maybe we can avoid this by asking Core Infra to provide us a way to calculate this. Let me discuss it internally

**Comment by ccorreagg** [comment]
Please @ZakiMiklos update the title with the correct LPD of the ticket. Now is "Lpd 44400" instead of "LPD-44400"

**Comment by ccorreagg** [comment]
Hey @ZakiMiklos ,

I have added a commit, as we don't need to calculate any future date. With the current date is enought.

Apart from that, I am reviewing the test. I am going to include comments

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java [hunk 2]
```diff
@@ -298,6 +298,43 @@ user, _getRandomDispatchExecutorType(),
 		}
 	}
 
+	@Test
+	public void testScheduleWithLaterCronExpressionSet() throws Exception {
```
Comment (ccorreagg) [inline modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:302]: This does not follow the pattern. The pattern must be test + <method we are testing> + other elements

For example: `testUpdateDispatchTriggerWithCronExpressions`

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java [hunk 3]
```diff
@@ -298,6 +298,43 @@ user, _getRandomDispatchExecutorType(),
...
+		DispatchTrigger dispatchTrigger = _addDispatchTrigger(
+			DispatchTriggerTestUtil.randomDispatchTrigger(
+				UserTestUtil.addUser(), _getRandomDispatchExecutorType(), 1));
+
+		DispatchTaskClusterMode dispatchTaskClusterMode =
+			DispatchTaskClusterMode.valueOf(
+				dispatchTrigger.getDispatchTaskClusterMode());
+
+		Calendar calendarStart = CalendarFactoryUtil.getCalendar(
```
Comment (ccorreagg) [inline modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:311]: This does not follow the pattern. In Liferay, we usually call the variables ending with the type, for example `startCalendar`

image

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java (hunk 2 above)
Comment (ccorreagg) [reply modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:302]: Furthermore, I would include different test cases, like:

// startDate in the past and after the cron expression

// startDate in the past and before the cron expression

// startDate in the future and before the cron expression

// startDate in the future and after the cron expression

// startDate null

What do you think?

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please review my comments

**Comment by ZakiMiklos** [comment]
ci:test:sf

**Comment by ZakiMiklos** [comment]
ci:test:relevant

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java [hunk 4]
```diff
@@ -455,6 +418,80 @@ public void testUpdateDispatchTriggerWhenMultiplePortalInstancesPresent()
...
+		Calendar nowCalendar = CalendarFactoryUtil.getCalendar(
+			new Date(
+			).getTime());
+
+		nowCalendar.add(Calendar.HOUR_OF_DAY, 12);
+
+		String cronExpression =
+			"0 0 " + nowCalendar.get(Calendar.HOUR_OF_DAY) + " * * ? *";
+
+		nowCalendar.add(0, 0);
```
Comment (ccorreagg) [inline modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:446]: What is this? Why are you not using the constant?

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java [hunk 5]
```diff
@@ -636,6 +673,34 @@ private String _getRandomDispatchExecutorType() {
...
+			dispatchTaskClusterMode, endCalendar.get(Calendar.MONTH),
+			endCalendar.get(Calendar.DAY_OF_MONTH),
+			endCalendar.get(Calendar.YEAR), 23, 59, false, true,
+			startCalendar.get(Calendar.MONTH),
+			startCalendar.get(Calendar.DAY_OF_MONTH),
+			startCalendar.get(Calendar.YEAR), 23, 59, "UTC");
+
+		Thread.sleep(1500);
+
+		Assert.assertEquals(
```
Comment (ccorreagg) [inline modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:698]: I don't understand this assert. For all the tests, we are asserting the same? That is weird

File: modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java (hunk 4 above)
Comment (ccorreagg) [reply modules/apps/dispatch/dispatch-test/src/testIntegration/java/com/liferay/dispatch/service/test/DispatchTriggerLocalServiceTest.java:446]: and adding a 0 value?

**Comment by ccorreagg** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:stop

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:forward

**Pull request closed without merging**